src/libs/kata-sys-util/src/cgroup.rs

```diff
@@ -0,0 +1,735 @@
+    disabled_hierarchies.clear();
+    for hierarchy in hierarchies {
+        //disabled_hierarchies.push(hie.clone());
+        match hierarchy.as_str() {
+            "blkio" => disabled_hierarchies.push(cgroups::Controllers::BlkIo),
+            "cpu" => disabled_hierarchies.push(cgroups::Controllers::Cpu),
+            "cpuset" => disabled_hierarchies.push(cgroups::Controllers::CpuSet),
+            "cpuacct" => disabled_hierarchies.push(cgroups::Controllers::CpuAcct),
+            "devices" => disabled_hierarchies.push(cgroups::Controllers::Devices),
+            "freezer" => disabled_hierarchies.push(cgroups::Controllers::Freezer),
+            "hugetlb" => disabled_hierarchies.push(cgroups::Controllers::HugeTlb),
+            "memory" => disabled_hierarchies.push(cgroups::Controllers::Mem),
+            "net_cls" => disabled_hierarchies.push(cgroups::Controllers::NetCls),
+            "net_prio" => disabled_hierarchies.push(cgroups::Controllers::NetPrio),
+            "perf_event" => disabled_hierarchies.push(cgroups::Controllers::PerfEvent),
+            "pids" => disabled_hierarchies.push(cgroups::Controllers::Pids),
+            "systemd" => disabled_hierarchies.push(cgroups::Controllers::Systemd),
+            _ => warn!(sl!(), "unknown cgroup controller {}", hierarchy),
+        }
+    }
+    debug!(
+        sl!(),
+        "disable cgroup list {:?} from {:?}", disabled_hierarchies, hierarchies
+    );
+}
+
+/// Filter out disabled cgroup subsystems.
+pub fn filter_disabled_cgroup(controllers: &mut Vec<Controllers>) {
+    let disabled_hierarchies = DISABLED_HIERARCHIES.lock().unwrap();
+    controllers.retain(|x| !disabled_hierarchies.contains(x));
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum PidType {
+    /// Add pid to `tasks`
+    Tasks,
+    /// Add pid to `cgroup.procs`
+    CgroupProcs,
+}
+
+/// Get the singleton instance for cgroup v1 hierarchy object.
+pub fn get_cgroup_hierarchies() -> &'static cgroups::hierarchies::V1 {
+    static GLOBAL: Lazy<cgroups::hierarchies::V1> = Lazy::new(cgroups::hierarchies::V1::new);
+    GLOBAL.deref()
+}
+
+// Prepend a kata specific string to oci cgroup path to form a different cgroup path, thus cAdvisor
+// couldn't find kata containers cgroup path on host to prevent it from grabbing the stats data.
+const CGROUP_KATA_PREFIX: &str = "kata";
+
+/// Convert to a Kata specific cgroup path.
+pub fn gen_kata_cgroup_path(path: &str) -> PathBuf {
+    // Be careful to trim off the possible '/' prefix. Joining an absolute path to a `Path` object
+    // will replace the old `Path` instead of concat.
+    Path::new(CGROUP_KATA_PREFIX).join(path.trim_start_matches('/'))
+}
+
+/// Convert to a cgroup path for K8S sandbox.
+pub fn gen_sandbox_cgroup_path(path: &str) -> PathBuf {
+    PathBuf::from(path)
+}
+
+/// A customized cgroup v1 hierarchy object with configurable filters for supported subsystems.
+#[derive(Debug)]
+pub struct V1Customized {
```

Comment:
This looks a lot like a subset of https://github.com/kata-containers/kata-containers/blob/main/src/agent/rustjail/src/cgroups and only works for cgroups v1? If so, please promote the agent implementation into a separate library and use that one for the runtime as well.